src/api-client.ts

```diff
@@ -0,0 +1,79 @@
+import * as msgpack from '@msgpack/msgpack';
+import { RpcRequest, RpcResponse } from '@osrf/romi-js-api-server/src/api-gateway';
+import Debug from 'debug';
+
+const debug = Debug('ApiClient');
+
+export default class ApiClient {
+  static async connect(url: string, token: string): Promise<ApiClient> {
+    const socket = new WebSocket(url);
+    socket.binaryType = 'arraybuffer';
+    await new Promise((res, reject) => {
+      const closeListener = (ev: CloseEvent) => {
+        reject(ev);
+      };
+      socket.addEventListener('close', closeListener);
+      socket.addEventListener('open', () => socket.send(token));
+      socket.addEventListener('message', (ev) => {
+        socket.removeEventListener('close', closeListener);
+        res(ev);
+      });
+    });
+    return new ApiClient(socket);
+  }
+
+  rpcRequest<Result = unknown, Param = unknown>(
+    method: string,
```

Comment:
That will be tricky to implement because the api gateway works in a plugin architecture (https://github.com/osrf/romi-js-api-server/tree/wip/src/plugins). We can do it for "in-tree" plugins, "out-of-tree" plugins would require some install time code generation.

An alternative is to do it this way
```ts
import type { pluginServices } from '@osrf/romi-js-api-server/src/plugins/ros2';

// where pluginServices is
export const pluginServices = {
  ros2Publish: 'ros2Publish';
}

// downside is that you don't really get intellisense
rpcRequest('string', ...);
// need to call it with
rpcRequest(pluginServices.ros2Publish, ...);
// but it still does not provide typings for the request and response types.
rpcRequest(pluginServices.ros2Publish, params /* unknown */, ...);
```

or better, by using generics

```ts
rpcRequest<Plugin, Method extends Plugin['methods']>(
  method: Method,
  params: Plugin[Method]['params'],
  ...
)

// where "Plugin" would be something like
export interface Ros2PluginMetadata {
  ros2Publish: {
    'params': PublishParams,
  }
  ...
}

// Downside of this is that you need to specify the plugin, but the advantage is that we will be able to get full typings.
import type { Ros2PluginMetadata} from '@osrf/romi-js-api-server/src/plugins/ros2'

rpcRequest<Ros2PluginMetadata>('ros2Publish', params /* PublishParams */, ...);
```

Reply:
> or better, by using generics
> 
> ```ts
> rpcRequest<Plugin, Method extends Plugin['methods']>(
>   method: Method,
>   params: Plugin[Method]['params'],
>   ...
> )
> 
> // where "Plugin" would be something like
> export interface Ros2PluginMetadata {
>   ros2Publish: {
>     'params': PublishParams,
>   }
>   ...
> }
> 
> // Downside of this is that you need to specify the plugin, but the advantage is that we will be able to get full typings.
> import type { Ros2PluginMetadata} from '@osrf/romi-js-api-server/src/plugins/ros2'
> 
> rpcRequest<Ros2PluginMetadata>('ros2Publish', params /* PublishParams */, ...);
> ```

Awesome! I think this will be better too. Since this is specifically a client for the `romi-js-api-server `  i think the developer should know or have access to the (future) docs of the plugin, so that shouldn't be a problem.